tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+that works in terms of not concrete types, but general combinators that
+describe shape and meta information about a data type. This way, we can
+declare how to perform a computation on almost any data type!
+
+## The idea about generics
+
+Haskell's mechanism that opens the possibility for such a thing as generics
+is, of course, type classes and ad-hoc polymorphism. The ability to describe
+a data type in terms of a set of combinators is our property, encapsulated
+by the `Generic` type class. Data type that is an instance of that type
+class can be processed by functions that work with representation of data
+type, not the data type itself. These functions, however, are by definition
+polymorphic and also usually hidden behind a type class interface, so the
+whole thing usually takes the form of automatic deriving of a type class
+instance.
+
+The process can be described like this:
+
+1. Given data type gets `Generic` instance automatically, as it can be
+   generated by the compiler with the help of `DeriveGeneric` language
+   extension.
+
+2. There is a type class instance defined for instances of `Generic` (or
+   `Generic1`). Since our data type is now an instance of those type
+   classes, that generics-based instance starts to work for us and we
+   magically get all the useful methods of that type class.
+
+If you know about `Data` and `Typeable` type classes, then you probably know
+that it's possible to do something similar using information which methods
+of those type classes provide, but since there are limits on the size of the
+tutorial, we won't touch this topic. You can start with
+this [blog post by Chris Done](http://chrisdone.com/posts/data-typeable) if
+you're interested.
+
+Step 2 is not so simple as it sounds though, at least not until we examine
+the details of this business more closely.
+
+## Combinators that represent shape of data type
+
+What could they look like? Well, if I showed you the data types as they are
+right now, you would probably run away in fear, cursing the tutorial and
+Haskell's generics. I have a better idea. Let's start with the simplest
+thing possible and then iterate asking ourselves how to work around some
+more interesting use cases that we would have to support. This will force us
+to make the data types more hairy but also more powerful step by step, till
+we arrive at the definitions that are actually used.
+
+For better or worse, algebraic data types lock us into a view on the world
+that consists of sums and products (another Haskell's feature that makes
+existence of generics framework possible, generics do not work with GADTs
+for example). So we need to be able to represent the following:
+
+* Data types without constructors at all: uninhabited types like `Void`.
+  This can be described as `data V1` (note: it has no constructors).
+
+* Constructors without arguments. Let this be something like this: `data U1
+  = U1`.
+
```

Comment:
I think it'll be obvious to anyone who is familiar with Haskell's data types. In short, sum data type means that you can have one of several options:

```haskell
data Foo = Foo | Bar | Baz
```

This type is inhibited by 3 values, represented by three *data constructors*.

To represent it generically we would need something like this:

```
Foo = L1 U1
Bar = R1 (L1 U1)
Baz = R1 (R1 U1)
```

You see, we need need this sum type to hold two data constructor “options” and then we can nest them and represent any sum type.

With product types, you multiple number of possible values that inhibit a type, i.e. with a tuple:

```haskell
(Foo, Foo)
```

Number of possible values is 3 × 3 = 9, because tuple includes one `Foo` value and another one. This is why this is called product types. This is why this is called *algebraic data types*. So with products you don't need two different constructors, because you have “everything at once”.